reactioncommerce/reaction
# Remove collection hooks for Products collection

Resolves #3634
Impact: Breaking
Type: Refactor

## Issue
Remove dependency on `matb33/meteor-collection-hooks` for the Products collection. This PR includes hooks related to Revisions. Further, there are additional related hooks for: Orders, Cart, see testing instructions for more details. 

## Solution
Refactor collection hooks to use Reactions hook system, and call explicitly were necessary. 

## Breaking changes

## Testing Instructions

**Revisions Before|After Insert Hook** - inserts a product revision into the `Revisions` collection before or after inserting a product into the `Products` collection.

Hooks Location: `imports/plugins/core/revisions/server/hooks.js`

Hook Names: `beforeInsertCatalogProductInsertRevision | afterInsertCatalogProductInsertRevision ` - perform a project search to find call sites

**Testing Procedures:**

User story 1: Create a new product, uses `afterInsertCatalogProductInsertRevision` hook

1. Create a new product
2. Verify a new revisions were inserted, one for the parent product `type=simple` and another one for `type=variant`
3. Verify new revision data was not published to `Products` collection
 

User story 2: Create a new varian on an existing product, uses `beforeInsertCatalogProductInsertRevision`

1. On an existing product create a new variant
2. Verify a new revisions was inserted for the new variant.

User story 3: Clone an existing product, uses `beforeInsertCatalogProductInsertRevision`

1. Clone a product
2. Verify corresponding revision were inserted.

User story 4: Clone an existing variant, uses `beforeInsertCatalogProductInsertRevision`

1. Clone a variant
2. Verify corresponding variant was inserted

___

**Revisions Before Update Hook** - updates a product revision in the `Revisions` collection before updating a product into the `Products` collection. The hook's return value determines whether the upderlying product should be updated.

Hooks Location: `imports/plugins/core/revisions/server/hooks.js`

Hook Names: `beforeUpdateCatalogProduct` - perform a project search to find call sites

User story 1: Update variant's position
1. On an existing product, have at least two variants created.
2. Rearrange variants positions
3. Verify position field: `index` was updated in the `Revisions` collection.
4. Publish product
5. Verify position field: `index` was updated in the `Products` collection

User story 2: Update a single product or variant field.
1. Update a product or variant field either: Title, price, visivility, vendor, description etc.
2. Verify field is updated in the corresponding revision.
3. Publish product
4. Verify field was updated in `Products` collection

User story 3: Update a product's tags
1. Update an existing product tag
2. Verify update in `Revisions`
3. Publish product
4. Verify update in `Products`

User story 4: Add a new tag
1. Add a new tag to a product
2. Verify new tag was adde to product's revision
3. Publish product
4. Verify new tag was added to `Products` collection.

User story 5: Remove tag
1. Remove a product's tag
2. Verify new tag was adde to product's revision
3. Publish product
4. Verify new tag was added to `Products` collection.

User story 6: Set new handle
1. Set a product's handle/permalink to a new value
2. Verify new handle in revisions
3. Publish product
4. Verify new handle in `Products` collection

User story 7: Update a product's position
1. Have at least two products created
2. Re-arrange their positions in the product grid
3. Verify changes are reflected in `Revisions` collection
4. "Publish All" products
5. Verify changes are reflected in `Products` collection

User story 8: Update an existing meta field
1. Update an existing metafield, A.K.A "Details"
2. Verify in update in revision, publish product, verify change in product record.

User story 9: Add new meta field
1. Add a new metafield
2. Verify in update in revision, publish product, verify change in product record.

User story 10: Remove new meta field
1. Remove a metafield
2. Verify in update in revision, publish product, verify change in product record.

User story 11: Toggle product visibility
1. Click on eye icon to toggle a product's visibility
2. Verify in update in revision, publish product, verify change in product record.

User story 11: Toggle variant visibility
1. Click on eye icon to toggle a variant's visibility
2. Verify in update in revision, publish product, verify change in product record.

___
**Revisions Before Remove Hook** - Marks a revision as deleted in the `Revisions` collection before updating a product into the `Products` collection. The hook's return value determines whether the upderlying product should be updated.

Hooks Location: `imports/plugins/core/revisions/server/hooks.js`

Hook Names: `beforeRemoveCatalogProduct` - perform a project search to find call sites

User story 1: Archive variant
1. Select a variant and click the archive button
2. Verify in update in revision, publish product, verify change in product record.

User story 2: Archive productc
1. Select a variant and click the archive button
2. Verify in update in revision, publish product, verify change in product record.
___
**Orders Before Update Hook** - Updates an order's workflow 

Hook Location: `/imports/plugins/core/orders/server/startup.js`

Hook Name: `beforeUpdateOrderWorkflow` - perform a project search to find call sites

User story 1: Update an orders workflow from "New" to "Processing" state
1. Place an order
2. Login as admin and click on blue arrow to change the orders state.
3. Verify order's workflow state transitioned from new to processing.

___
**Cart After Update Hook** - Re-calculates a cart's taxes after an update to the cart.

Hook Location: `/imports/plugins/core/taxes/server/hooks/taxes.js`

Hook Name: `afterCartUpdateCalculateTaxes"` - perform a project search to find call sites

NOTE: Enable taxes for all user stories.
User story 1: Add an item to cart 
1. Add a new product to the shopping cart.
4. Verify taxes are calculated correctly in the shopping cart.

User story 2: Remove an item to cart 
1. Remove a product to the shopping cart.
4. Verify taxes are calculated correctly in the shopping cart.

User story 3: Enter a shipping address
1. Add a product to the cart
2. Start checkout process
4. Enter shipping address
5. Verify taxes calculation is corrects

User story 5: Set shipping address, then remove.
1. Add a product to the cart
2. Start checkout process
4. Enter shipping address
5. Remove shipping address
5. Verify taxes re-calculation in triggered, (hook is ran)

User story 6: Set payment method
1. Add a product to the cart
2. Start checkout process
4. Select a payment method
5. Verify taxes calculation is correct for the payment method

User story 7: Remove payment method
1. Add a product to the cart
2. Start checkout process
4. Select a payment method
5. Select new payment method
5. Verify taxes re-calculation is triggered(hook is ran)

**Publish Order after intert Hook** - Publishes newly created orders to shopify, if hook is enabled in shopify connector

Hook Location: `imports/plugins/included/connectors-shopify/server/hooks/orders.js` 

Hook Name: `afterOrderInsert"` - perform a project search to find call sites

User story 1: Create a new order
1. Enable "Push orders to Shopify on create" setting in shopify connector
2. Create a order
3. Verify a job was created to export the newly created order to shopify.
